src/ray/core_worker/actor_manager.cc

```diff
@@ -299,5 +273,69 @@ ActorID ActorManager::GetCachedNamedActorID(const std::string &actor_name) {
   return ActorID::Nil();
 }
 
+void ActorManager::SubscribeActorState(const ActorID &actor_id) {
+  {
+    // Make sure this method only be executed once.
+    absl::MutexLock lock(&subscription_mutex_);
+    if (!subscribed_actors_.emplace(actor_id, /*valid*/ true).second) {
+      return;
+    }
+  }
+
+  auto actor_handle = GetActorHandle(actor_id);
+  RAY_CHECK(actor_handle != nullptr);
+  auto cached_actor_name =
+      GenerateCachedActorName(actor_handle->GetNamespace(), actor_handle->GetName());
+
+  // Register a callback to handle actor notifications.
+  auto actor_notification_callback =
+      std::bind(&ActorManager::HandleActorStateNotification,
+                this,
+                std::placeholders::_1,
+                std::placeholders::_2);
+  RAY_CHECK_OK(gcs_client_->Actors().AsyncSubscribe(
+      actor_id,
+      actor_notification_callback,
+      [this, actor_id, cached_actor_name](Status status) {
+        if (status.ok() && !cached_actor_name.empty()) {
+          // It should be ignored if the finished callback of the subscription comes
+          // after the actor died, otherwise the stale named actor will be cached.
+          // NOTE: We can not guarantee the order of arrival of `on_done` callback and
+          // subscribe callback for the time being.
+          if (IsValidActor(actor_id)) {
+            absl::MutexLock lock(&cache_mutex_);
+            cached_actor_name_to_ids_.emplace(cached_actor_name, actor_id);
+          }
```

Comment:
Should we remove it from `subscribed_actors_` if it's there and dead?

Reply:
> subscribed_actors_

I'm afraid not, beacuse the `subscribed_actors_`  is used to guarantee that the subscription of actor is only invoked once, if we remove the invalid actor from `subscribed_actors_`, one can use an stale ActorHandle to submit actor task and resubscribe again.

Reply:
This is kind of wired, does this means as time goes, we'll have cached_actor_name_to_ids_ accumulates forever unless we restart the raylet?

Reply:
> This is kind of wired, does this means as time goes, we'll have cached_actor_name_to_ids_ accumulates forever unless we restart the raylet?

The named actor will be removed from `cached_actor_name_to_ids_` when the actor is invalid, while the entity of `subscribed_actors_` will not do that.

BTW, no actor_handle will be erased from `actor_handles_` for the time being(I searched all the codes, there is no `actor_handles_.erase`), so I just make `cached_actor_name_to_ids_` the same as `actor_handles_`. Both `actor_handles_` and `cached_actor_name_to_ids_` are inside `CoreWorker`(not `Raylet`), so I think it's ok for now?

Reply:
I see. I'm good with this then.